### ISSUE TITLE
New transport protocol (sunsetting JSON)

### ISSUE THREAD

**Issue by Saiv46** [matrix-org/matrix-spec#1736]
**Problem**
The Client-Server API is not bandwidth efficient enough. matrix-org/matrix-spec-proposals#3079 was supposed to help with that, but it has too broad scope and has the own issues. One of them is E2E — despite having CBOR as data format, encrypted messages are represented as Base64 string.

**Suggestion**
I suggest allowing client and server to specify request and response media types as by [RFC 7231 Section 5.3.2](https://datatracker.ietf.org/doc/html/rfc7231#section-5.3.2). This does **not** affect HTTP API and shouldn't require such changes.

Client and servers must set supported formats in `Accept` and `Content-Type` accordingly.
```
Accept: application/msgpack, example/vnd.elementx, application/json
Content-Type: application/msgpack
```
In this example I'm using [MessagePack](https://msgpack.org/) instead of CBOR, because its encoding is less complex (~~and wasn't a product of IETF bikeshedding~~), so it would provide better performance with the same bandwidth efficiency.

After some time, new proposal can be made to switch underlying E2E data encoding to most efficient one that are already being used, but only if MSC3079 would be useful enough to justify such changes.

**Comment by richvdh** [comment]
Related: https://github.com/matrix-org/matrix-spec/issues/1460

**Comment by kegsay** [comment]
I doubt JSON would _ever_ be "sunsetted", it is too ubiquitous, human-readable and easy to use.

MessagePack and CBOR encodings are extremely similar, given they started out as one proposal before diverging. For "common" types they are _effectively_ interchangable. For bandwidth optimisation, both are as efficient as each other. Encode the following and you'll see they are both 131 bytes:
```
{
  “name”: “Alice”,
  “age”: 42,
  “refs”: [11, 65540, null, 4294967300],
  “friends”: {
    “Bob”: “AABBCCDDEEFFGG”,
    “Carol”:” “HHIIJJKKLLMMNNOO” 
  }
  “x”: 0.1,
  “y”: null,
  “valid”: true
}
```
As for performance, I think it's not true to state "use MessagePack because its encoding is less complex, so it would provide better performance with the same bandwidth efficiency." Most benchmarks like https://prataprc.github.io/msgpack-vs-cbor.html show they are broadly equivalent, where most of the performance is due to the quality of the library rather than the protocol itself. Because there is so little to differentiate them, CBOR at least has a standardised specification as an RFC vs MessagePack which has a specification in [github](https://github.com/msgpack/msgpack/blob/master/spec.md).

The problem with saying "just use MessagePack/CBOR" is that is actually _doesn't compress very well_ when you're dealing with string keys and string values, because the strings are basically left alone when encoding. There's a reason why MSC3079 uses enum integer keys. Look at those 131 bytes and you'll see a lot is due to key names, which you can compress significantly if you swapped them out for integers. The longer the key name, the greater the benefit, and Matrix JSON objects often have long key names. You can't go all the way and omit the key names entirely ala Protocol Buffers either, because Matrix events need to be extensible (so you can't add keys not known ahead of time). Protocol Buffers does support extensible keys, but it's clunky (often just falling back to using JSON as a serialisation format) and you lose a lot of the things you would use Protocol Buffers for in the first place like static type checking.


tl;dr converting from JSON to MessagePack won't get you as much as you think, you need additional tweaks for it to be worthwhile, which is why MSC3079 exists. You could feasibly split up MSC3079 into "CBOR bits" and "everything else" and use the Accept header, which would definitely be a possible stepping stone for low bandwidth support in Matrix.

**Comment by Saiv46** [comment]
Utilizing CBOR tags and specification constraints we can compress more things:

- Encrypted data: **33%** reduction by encoding from base64
- Event ID: **49%** reduction by storing reference hash directly, and **25%** for opaque strings
- User/Room ID and aliases: **25%** reduction by treating string as base43
- CBOR integer keys from [MSC3079](https://github.com/matrix-org/matrix-spec-proposals/pull/3079) was NOT implemented

Wrote a little script to test with sampled requests from Element Web
```
┌───────────────┬──────────────────────┐
│ (index)       │ Values               │
├───────────────┼──────────────────────┤
│ Original JSON │ '493835B (baseline)' │
│ UTF-8 JSON    │ '413490B (-16%)'     │
│ Plain CBOR    │ '373990B (-24%)'     │
│ Tagged CBOR   │ '362992B (-26%)'     │
└───────────────┴──────────────────────┘
```

Just adding `Content-Type: application/json; charset=utf-8` header and disabling unicode escaping will result in 16% bandwidth reduction, but from there we have diminishing returns as messages can't be compressed either due to security implications or because they're encrypted.

**Comment by rafael2k** [comment]
For the low bitrate use case, I'd also apply some type of compression, like huffman, deflate or arithmetic coding, on top of  the MessagePack (this is before or after encryption?).

**Comment by Saiv46** [comment]
> For the low bitrate use case, I'd also apply some type of compression, like huffman, deflate or arithmetic coding, on top of the MessagePack (this is before or after encryption?).

Compression before encryption has a security flaw (also there isn't much repeating characters), compression after encryption is useless.

**Comment by Saiv46** [comment]
Closed in favor matrix-org/matrix-spec-proposals#4224

**Comment by rafael2k** [comment]
As I understand, each character is occupying one byte (or more), right? Considering compression before encryption, it would still be possible to have relevant compression, right?
Btw, which kind of security flaw you mean by applying compression before encryption?
Thanks @Saiv46!

**Comment by Saiv46** [comment]
> As I understand, each character is occupying one byte (or more), right? Considering compression before encryption, it would still be possible to have relevant compression, right?
> Btw, which kind of security flaw you mean by applying compression before encryption?
> Thanks @Saiv46!

Hi again @rafael2k.

Compression relies on the lack of entropy (redundancy), while encryption maximises it to obscure its content. The security flaw I mentioned is attacks like BREACH, which is made possible by observing compressed size contents.

**Comment by rafael2k** [comment]
Hi @Saiv46, sorry to bother.
: )

Considering that most of the times there is redundancy, unless you use optimal code for each symbols, such vulnerabilities do not rule out compression, for example, in an environment when compression is relevant for the service to exist in the first place.

**Comment by Saiv46** [comment]
Good evening @rafael2k !

> Considering that most of the times there is redundancy, unless you use optimal code for each symbols, such vulnerabilities do not rule out compression, for example, in an environment when compression is relevant for the service to exist in the first place.

What do you mean by such environment? CBOR encoding have a goal of using optimal code for each value, which makes it susceptible to similar vulnerabilities.

Of course there's mitigations like length hiding, separating secrets from user input, masking secrets, and/or rate-limiting requests. But that's gonna be a topic for next proposals.

**Comment by clokep** [comment]
> Closed in favor [matrix-org/matrix-spec-proposals#4224](https://github.com/matrix-org/matrix-spec-proposals/pull/4224)

Let's keep this open until that lands as it is possible for that MSC to be rejected and replaced by a different one.

**Comment by rafael2k** [comment]
Morning @Saiv46!

> What do you mean by such environment? CBOR encoding have a goal of using optimal code for each value, which makes it susceptible to similar vulnerabilities.

Right. Got it. Btw, I mean, a network environment very bit-rate constrained, in which any compression is very valuable

> Of course there's mitigations like length hiding, separating secrets from user input, masking secrets, and/or rate-limiting requests. But that's gonna be a topic for next proposals.

Right. Thanks for re-opening.

**Comment by TheArcaneBrony** [comment]
I agree that using compression is most likely far more worthwhile, as someone who's living at 12 Mbps :)
Find the compression options for your nearest reverse proxy (zstd is cool btw, just unsupported by safari), and apply it to the most size-heavy endpoints (members, state, sync...), or all of them.
